dali/kernels/span.h

```diff
@@ -0,0 +1,145 @@
+  constexpr bool empty() const noexcept { return false; }
+
+  // [span.elem], span element access
+  constexpr reference operator[](index_type idx) const { return data_[idx]; }
+  constexpr reference operator()(index_type idx) const { return data_[idx]; }
+  constexpr pointer data() const noexcept { return data_; }
+
+  // [span.iterators], span iterator support
+  constexpr iterator begin() const noexcept { return data_; }
+  constexpr iterator end() const noexcept { return data_ + Extent; }
+  constexpr const_iterator cbegin() const noexcept { return data_; }
+  constexpr const_iterator cend() const noexcept { return data_ + Extent; }
+
+ private:
+  pointer data_;
+};
+
+template <class ElementType>
+class span<ElementType, dynamic_extent> {
+ public:
+  // constants and types
+  using element_type = ElementType;
+  using value_type = typename std::remove_cv<ElementType>::type;
+  using index_type = ptrdiff_t;
+  using difference_type = ptrdiff_t;
+  using pointer = element_type *;
+  using reference = element_type &;
+  using iterator = element_type *;
+  using const_iterator = const element_type *;
+  static constexpr index_type extent = dynamic_extent;
+
+  // [span.cons], span constructors, copy, assignment, and destructor
+  constexpr span() noexcept : data_(nullptr), size_(0) {}
+  constexpr span(pointer ptr, index_type count) : data_(ptr), size_(count) {}
+  constexpr span(pointer firstElem, pointer lastElem) : data_(firstElem), size_(lastElem - firstElem) {}
+  constexpr span(const span &other) noexcept = default;
+  ~span() noexcept = default;
+  span &operator=(const span &other) noexcept = default;
+
+  // [span.sub], span subviews
+
+  // [span.obs], span observers
+  constexpr index_type size() const noexcept { return size_; }
+  constexpr index_type size_bytes() const noexcept { return size_ * sizeof(value_type); }
+  constexpr bool empty() const noexcept { return size() == 0; }
+
+  // [span.elem], span element access
+  constexpr reference operator[](index_type idx) const { return data_[idx]; }
+  constexpr reference operator()(index_type idx) const { return data_[idx]; }
+  constexpr pointer data() const noexcept { return data_; }
+
+  // [span.iterators], span iterator support
+  constexpr iterator begin() const noexcept { return data_; }
+  constexpr iterator end() const noexcept { return data_ + size(); }
+  constexpr const_iterator cbegin() const noexcept { return data_; }
+  constexpr const_iterator cend() const noexcept { return data_ + size(); }
+
+ private:
+  pointer data_;
+  index_type size_;
+};
+
+
+// [span.comparison], span comparison operators
+template <class ElementL, ptrdiff_t ExtentL, class ElementR, ptrdiff_t ExtentR>
+/* constexpr */bool operator==(span<ElementL, ExtentL> l, span<ElementR, ExtentR> r) {
+  if (l.size() != r.size()) {
+    return false;
+  }
+  auto size = l.size();
+  for (size_t i = 0; i < size; i++) {
```

Comment:
This will cause a compiler warning: "signed/unsigned comparison". Use `decltype(size) i`.